libflux/src/semantic/tests.rs

```diff
@@ -0,0 +1,183 @@
+use std::collections::HashMap;
+
+use crate::semantic::analyze::analyze;
+use crate::semantic::env::Environment;
+use crate::semantic::fresh::Fresher;
+use crate::semantic::nodes;
+use crate::semantic::parser::parse;
+use crate::semantic::types::PolyType;
+
+use crate::ast;
+use crate::parser::parse_string;
+
+#[cfg(test)]
+use colored::*;
+
+fn parse_program(src: &str) -> ast::Package {
+    let file = parse_string("program", src);
+
+    ast::Package {
+        base: file.base.clone(),
+        path: "path".to_string(),
+        package: "main".to_string(),
+        files: vec![file],
+    }
+}
+
+/// The test_infer! macro generates test cases for type inference.
+///
+/// A test case consists of:
+///
+/// 1. A name given as an identifier
+/// 2. An optional type environment (representing a prelude)
+/// 3. The flux program to be type checked
+/// 4. The expected types of the top-level identifiers of said program
+///
+/// # Example
+///
+/// ```
+/// test_infer! { instantiation,
+///     env: &[
+///         ("f", "forall [t0] where t0: Addable (a: t0, b: t0) -> t0"),
+///     ],
+///     src: "x = f",
+///     exp: &[
+///         ("x", "forall [t0] where t0: Addable (a: t0, b: t0) -> t0"),
+///     ],
+/// }
+/// ```
+///
+macro_rules! test_infer {
+    (
+        $name:ident,
+        $(env: $env:expr,)?
+        src: $src:expr,
+        exp: $exp:expr,
+    ) => {
+        #[test]
+        fn $name() {
```

Comment:
So just a quick comment, but can we have it so `test_infer` is called from within a function? I'm not too keen on the macro generating the function itself. The entirety of the body seems fine. It's just I'd like:

```
fn test_blah_blah() {
    test_infer!()
}
```

or whatever the proper syntax would be.